buildSrc/src/main/java/com/google/firebase/gradle/ReleaseGenerator.kt

```diff
@@ -0,0 +1,118 @@
+package com.google.firebase.gradle
+
+import java.io.File
+import java.io.IOException
+import org.eclipse.jgit.api.Git
+import org.eclipse.jgit.api.ListBranchCommand
+import org.eclipse.jgit.api.errors.GitAPIException
+import org.eclipse.jgit.lib.Constants
+import org.eclipse.jgit.lib.ObjectId
+import org.eclipse.jgit.lib.Ref
+import org.gradle.api.DefaultTask
+import org.gradle.api.tasks.TaskAction
+
+open class ReleaseGenerator : DefaultTask() {
+    @TaskAction
+    @Throws(Exception::class)
+    fun generateReleaseConfig() {
+        val currentRelease = project.property("currentRelease").toString()
+        val pastRelease = project.property("pastRelease").toString()
+        val rootDir = project.rootDir
+        val availableModules = parseSubProjects(rootDir)
+
+        val repo = initRepo(rootDir)
+        val headRef = repo.repository.resolve(Constants.HEAD)
+        val branchRef = getObjectRefForBranchName(repo, pastRelease)
+
+        val changedDirs = getChangedLibraries(repo, branchRef, headRef, FIREBASE_LIBRARIES)
+        val changedModules = findChangedModules(changedDirs, availableModules)
+        writeReleaseConfig(rootDir, changedModules, currentRelease)
+    }
+
+    private fun findChangedModules(changedDirs: List<String>, availableModules: Set<String>): List<String> = changedDirs.flatMap { directory ->
+        val moduleName = directory.replace("/", ":")
+        val ktxModuleName = "$moduleName:ktx"
+        listOf(moduleName, ktxModuleName).filter { availableModules.contains(it) }
+    }
+
+    private fun parseSubProjects(rootDir: File) = File(rootDir, "subprojects.cfg").readLines().filter { !it.startsWith("#") }.toSet()
+
+    @Throws(IOException::class)
+    private fun initRepo(repoPath: File): Git {
+        if (!repoPath.exists()) {
+            throw RuntimeException("Repo path doesn't exist")
+        }
+        return Git.open(repoPath)
+    }
+
+    @Throws(GitAPIException::class)
+    private fun getObjectRefForBranchName(repo: Git, branchName: String): ObjectId {
+        return repo.branchList().setListMode(ListBranchCommand.ListMode.REMOTE).call()
+                .filter { ref: Ref -> ref.name == "refs/remotes/origin/$branchName" }
+                .map { obj: Ref -> obj.objectId }.firstOrNull()
+                ?: throw RuntimeException("Could not find branch named $branchName")
+    }
+
+    private fun getChangedLibraries(repo: Git, startRef: ObjectId, endRef: ObjectId, directories: List<String>): List<String> {
+        return FIREBASE_LIBRARIES.filter { libraryPath: String ->
+            repo.log()
+                    .addPath("$libraryPath/*.kt")
```

Comment:
Doesn't look like we can get away with this, unless I'm mistaken:
```
Note that regex expressions or wildcards are not supported.
```
From the docs for `addPath`

Reply:
I hadn't noticed that, I was following the more generic git documentation. 

My gut is that despite what the documentation claims, this form of wildcard (similar to the git log command) does work.

because interpreted literally, there is no diff that could include this.

Reply:
i've changed this to just the directory to resolve a bug noticed by rlazo below. I'll try to narrow it in from here.